README.md
# Web Bank


A simple web application in Clojure for managing bank accounts.

## Installation

This project uses **Leiningen** to build and manage its dependencies. Before continuing, make sure you have [Java](https://www.oracle.com/technetwork/java/javase/downloads/index.html) version 1.6 or later installed. You can check your Java version by running the command:

    $ java -version


Then install Leiningen using the instructions on [Leiningen home page](https://leiningen.org/).

The code here presented has been developed and tested on Linux (Ubuntu 14.04), with Leiningen 2.7.1 on JRE 1.8 version.


## Usage

After cloning the repository, first navigate to the root project directory and download all project dependencies.

    $ lein deps

After which you can run the server locally.

    $ lein run

The application runs on port 5000. So any POST requests must be made to *localhost:5000*.


## Routes

All HTTP endpoints accept and return JSON payloads as requests and responses.

Endpoint   | Bank Operation
---------- | ------------------
/transactions | Adds a credit or debit transaction to a given account
/balances | Gets current balance of a given account
/statements | Gets bank account statement for a given period
/debts | Gets the periods where balances of given account has been negative


## Requests

[Here](https://github.com/DanLux/nubank/tree/master/resources/requests-examples) you have json examples for all requests available.

For a credit<sup>[1](https://github.com/DanLux/nubank/blob/master/resources/requests-examples/credit-transactions.json)</sup> or debit<sup>[1'](https://github.com/DanLux/nubank/blob/master/resources/requests-examples/debit-transactions.json)</sup> transaction, current account balance<sup>[2](https://github.com/DanLux/nubank/blob/master/resources/requests-examples/balances.json)</sup>, account statement<sup>[3](https://github.com/DanLux/nubank/blob/master/resources/requests-examples/statements.json)</sup> and periods of debt<sup>[4](https://github.com/DanLux/nubank/blob/master/resources/requests-examples/debts.json)</sup>.

> All values in the examples are arbitrary.
One must only pay attention to the json string keys and the json value types and formats.


## Responses

[Here](https://github.com/DanLux/nubank/tree/master/resources/responses-examples) you have json examples for the expected responses.

For a credit or debit transaction<sup>[1](https://github.com/DanLux/nubank/blob/master/resources/responses-examples/transactions.json)</sup>, current account balance<sup>[2](https://github.com/DanLux/nubank/blob/master/resources/responses-examples/balances.json)</sup>, account statement<sup>[3](https://github.com/DanLux/nubank/blob/master/resources/responses-examples/statements.json)</sup> and periods of debt<sup>[4](https://github.com/DanLux/nubank/blob/master/resources/responses-examples/debts.json)</sup>.


## Testing

You can run all project tests with the command:

    $ lein test

And a report with all results will be automatically displayed.


## License

Distributed under the Eclipse Public License.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-10-07
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-10-07
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/nubank/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/nubank/compare/0.1.0...0.1.1


test/webbank/bank_test.clj
(ns webbank.bank-test
	(:require
		[clojure.test :refer :all]
		[webbank.date-converter :as converter]
		[webbank.bank :refer :all]))


(deftest test-get-account-by
	(let [account-number "ABC01678"]
		(testing "when there is no such account"
			(reset-bank)
			(is (nil? (get @bank account-number)))
			(is (zero? (count (vals @bank))))
			(is (= (sorted-map) (get-account-by account-number)))
			(is (= 1 (count (vals @bank))))
		)

		(testing "when there is such account"
			(is ((complement nil?) (get @bank account-number)))
			(is (= 1 (count (vals @bank))))
			(is (= (get @bank account-number) (get-account-by account-number)))
		)
	)
)

(deftest test-add-account-transaction
	(testing "trying to add transactions with zero amount"
		(let [account-number "000000-00"
			description "Doing nothing"
			amount 0.00M]

			(reset-bank)
			(add-account-transaction account-number description amount (converter/from-string "2017-04-01"))
			(add-account-transaction account-number description amount (converter/from-string "2017-04-02"))
			(add-account-transaction account-number description amount (converter/from-string "2017-04-03"))
			(is (empty? (get @bank account-number)))
		)
	)

	(testing "adding a single credit transaction"
		(let [account-number "ABC01678"
			description "Deposit"
			amount 1405.00M
			date (converter/from-string "2017-08-22")
			expected-transaction {:description description, :amount amount}]

			(reset-bank)
			(add-account-transaction account-number description amount date)
			(is (= 1 (count (vals @bank))))
			(is (every? (partial = expected-transaction) (get-in @bank [account-number date])))
		)
	)

	(testing "adding a single debit transaction"
		(let [account-number "GTA961678-X"
			description "Withdrawal"
			amount -760.75M
			date (converter/from-string "2017-09-01")
			expected-transaction {:description description, :amount amount}]

			(reset-bank)
			(add-account-transaction account-number description amount date)
			(is (= 1 (count (vals @bank))))
			(is (every? (partial = expected-transaction) (get-in @bank [account-number date])))
		)
	)

	(testing "adding multiple transactions on same day"
		(let [account-number "GTA961678-X"
			credit-description "Withdrawal"
			debit-description "Deposit"
			credit-amount 245.74M
			debit-amount -170.21M
			date (converter/from-string "2017-04-03")
			expected-credit-transaction {:description credit-description, :amount credit-amount}
			expected-debit-transaction {:description debit-description, :amount debit-amount}]

			(reset-bank)
			(add-account-transaction account-number credit-description credit-amount date)
			(add-account-transaction account-number debit-description debit-amount date)
			(is (= 1 (count (vals @bank))))
			(is (some (partial = expected-credit-transaction) (get-in @bank [account-number date])))
			(is (some (partial = expected-debit-transaction) (get-in @bank [account-number date])))
			(is (= 2 (count (get-in @bank [account-number date]))))
		)
	)

	(testing "adding multiple transactions on different days and keeping them sorted"
		(let [account-number "00007"
			feb-date (converter/from-string "2017-02-21")
			jul-date (converter/from-string "2017-07-14")
			dec-date (converter/from-string "2017-12-31")
			expected-feb-transaction {:description "February salary", :amount 6500.00M}
			expected-jul-transaction {:description "July vacation", :amount -2604.01M}
			expected-dec-transaction {:description "New year purchases", :amount -590.49M}]

			(reset-bank)
			(add-account-transaction account-number "New year purchases" -590.49M dec-date)
			(add-account-transaction account-number "February salary" 6500.00M feb-date)
			(add-account-transaction account-number "July vacation" -2604.01M jul-date)
			(is (= 1 (count (vals @bank))))
			(is (every? (partial = expected-feb-transaction) (get-in @bank [account-number feb-date])))
			(is (every? (partial = expected-jul-transaction) (get-in @bank [account-number jul-date])))
			(is (every? (partial = expected-dec-transaction) (get-in @bank [account-number dec-date])))
			(is (= feb-date (first (keys (get @bank account-number)))))
			(is (= jul-date (second (keys (get @bank account-number)))))
			(is (= dec-date (last (keys (get @bank account-number)))))
		)
	)
)

(deftest test-balances-and-statements
	(let [account-number "BCA22156-0"
		just-made-account-number "99999XYZ-0"
		nonexistent-account-number "XXX-NOT-FOUND-XXX"
		before-yesterday (converter/from-string "2017-10-15")
		yesterday (converter/from-string "2017-10-16")
		today (converter/from-string "2017-10-17")
		tomorrow (converter/from-string "2017-10-18")]

		(reset-bank)
		(get-account-by just-made-account-number)
		(add-account-transaction account-number "Purchase on Amazon" -3.34M yesterday)
		(add-account-transaction account-number "Withdrawal" -180.00M today)
		(add-account-transaction account-number "Purchase on Uber" -45.23M yesterday)
		(add-account-transaction account-number "Deposit" 1000.00M before-yesterday)

		(testing "daily-transactions-balance"
			(testing "when no transactions were made on given date"
				(is (zero? (daily-transactions-balance (get @bank account-number) tomorrow))))

			(testing "when transactions were made on given date"
				(is (= 1000.00M (daily-transactions-balance (get @bank account-number) before-yesterday)))
				(is (= -48.57M (daily-transactions-balance (get @bank account-number) yesterday)))
				(is (= -180.00M (daily-transactions-balance (get @bank account-number) today))))
		)

		(testing "current-balance"
			(testing "when there is no such account"
				(is (zero? (current-balance nonexistent-account-number))))

			(testing "when account has just been created"
				(is (zero? (current-balance just-made-account-number))))

			(testing "for a regular used account"
				(is (= 771.43M (current-balance account-number))))
		)

		(testing "account-statement"
			(testing "when there is no such account"
				(is (= (sorted-map) (account-statement nonexistent-account-number))))

			(testing "when account has just been created"
				(is (= (sorted-map) (account-statement just-made-account-number))))

			(testing "for a regular used account"
				(testing "without specific period"
					(let [expected-statement {
						before-yesterday [
							[{:description "Deposit", :amount 1000.00M}],
							1000.00M
						],

						yesterday [
							[{:description "Purchase on Amazon", :amount -3.34M}, {:description "Purchase on Uber", :amount -45.23M}],
							951.43M
						],

						today [
							[{:description "Withdrawal", :amount -180.00M}],
							771.43M
						]
					}]

					(is (= expected-statement (account-statement account-number))))
				)

				(testing "given an early start date and past end date"
					(let [expected-statement {
						before-yesterday [
							[{:description "Deposit", :amount 1000.00M}],
							1000.00M
						],

						yesterday [
							[{:description "Purchase on Amazon", :amount -3.34M}, {:description "Purchase on Uber", :amount -45.23M}],
							951.43M
						]
					}]

					(is (= expected-statement (account-statement account-number (converter/from-string "2010-08-26") yesterday)))
					(is (= expected-statement (account-statement account-number (converter/from-string "2017-10-15") yesterday)))
					(is (= expected-statement (account-statement account-number before-yesterday yesterday))))
				)

				(testing "given an late start date and future end date"
					(let [expected-statement {
						yesterday [
							[{:description "Purchase on Amazon", :amount -3.34M}, {:description "Purchase on Uber", :amount -45.23M}],
							951.43M
						],

						today [
							[{:description "Withdrawal", :amount -180.00M}],
							771.43M
						]
					}]

					(is (= expected-statement (account-statement account-number yesterday tomorrow)))
					(is (= expected-statement (account-statement account-number yesterday today))))
				)

				(testing "multi-arity equivalence"
					(is (= (account-statement account-number) (account-statement account-number before-yesterday today)))
				)
			)
		)
	)
)

(deftest test-periods-of-debt
	(let [account-number "1000000000000066600000000000001"
		nonexistent-account-number "XXX-NOT-FOUND-XXX"
		just-made-account-number "0000004894-Y"]

		(reset-bank)
		(get-account-by just-made-account-number)

		(testing "when there is no such account"
			(is (= [] (periods-of-debt nonexistent-account-number))))

		(testing "when account has just been created"
			(is (= [] (periods-of-debt just-made-account-number))))

		(testing "for a regular used account"
			(testing "when there was never a single period of debt"
				(add-account-transaction account-number "Deposit" 1000.00M (converter/from-string "2017-10-15"))
				(add-account-transaction account-number "Purchase on Amazon" -3.34M (converter/from-string "2017-10-16"))
				(add-account-transaction account-number "Purchase on Uber" -45.23M (converter/from-string "2017-10-16"))
				(add-account-transaction account-number "Withdrawal" -180.00M (converter/from-string "2017-10-17"))

				(is (= [] (periods-of-debt account-number)))
			)

			(testing "when there is a current period of debt"
				(add-account-transaction account-number "Purchase of a flight ticket" -800.00M (converter/from-string "2017-10-18"))
				(add-account-transaction account-number "Purchase of a espresso" -10.00M (converter/from-string "2017-10-22"))

				(let [expected-period [{:principal 28.57M, :start "2017-10-18", :end "2017-10-21"},
									{:principal 38.57M, :start "2017-10-22"}]]
				(is (= expected-period (periods-of-debt account-number))))
			)

			(testing "when there is no current period of debt"
				(add-account-transaction account-number "Deposit" 38.57M (converter/from-string "2017-10-25"))

				(let [expected-period [{:principal 28.57M, :start "2017-10-18", :end "2017-10-21"},
									{:principal 38.57M, :start "2017-10-22", :end "2017-10-24"}]]
				(is (= expected-period (periods-of-debt account-number))))
			)
		)
	)
)


test/webbank/core_test.clj
(ns webbank.core-test
	(:require
		[clojure.test :refer :all]
		[ring.mock.request :as mock]
		[cheshire.core :refer :all]
		[webbank.core :as core]))

(defn request-application
	[endpoint json]
	(-> (mock/request :post endpoint json)
		(mock/content-type "application/json")
		(core/application)
	)
)

(deftest test-routes
	(testing "/transactions endpoint"
		(testing "for a credit transaction"
			(let [json-request (generate-string { "account" "BA47856", "description" "Salary", "amount" 2371.20M, "date" "2016-07-13"})
				expected-response (generate-string {})
				response (request-application "/transactions" json-request)]

				(is (= 200 (:status response)))
				(is (= "application/json; charset=utf-8" (get-in response [:headers "Content-Type"])))
				(is (= expected-response (:body response)))
			)
		)

		(testing "for a debit transaction"
			(let [json-request (generate-string { "account" "BA47856", "description" "Purchase", "amount" -3651.98M, "date" "2015-12-21"})
				expected-response (generate-string {})
				response (request-application "/transactions" json-request)]

				(is (= 200 (:status response)))
				(is (= "application/json; charset=utf-8" (get-in response [:headers "Content-Type"])))
				(is (= expected-response (:body response)))
			)
		)
	)

	(testing "/balances endpoint"
		(let [json-request (generate-string {"account" "BA47856"})
			expected-response (generate-string {"amount" -1280.78M})
			response (request-application "/balances" json-request)]

			(is (= 200 (:status response)))
			(is (= "application/json; charset=utf-8" (get-in response [:headers "Content-Type"])))
			(is (= expected-response (:body response)))
		)
	)

	(testing "/statements endpoint"
		(let [json-request (generate-string {"account" "BA47856", "start" "2015-12-21", "end" "2016-07-13"})
			expected-response (generate-string {
				"2015-12-21" [[{"description" "Purchase", "amount" -3651.98M}], -3651.98M],
				"2016-07-13" [[{"description" "Salary", "amount" 2371.20M}], -1280.78M]
			})
			response (request-application "/statements" json-request)]

			(is (= 200 (:status response)))
			(is (= "application/json; charset=utf-8" (get-in response [:headers "Content-Type"])))
			(is (= expected-response (:body response)))
		)
	)

	(testing "/debts endpoint"
		(let [json-request (generate-string {"account" "BA47856"})
			expected-response (generate-string {
				"debts" [{"principal" 3651.98M, "start" "2015-12-21", "end" "2016-07-12"},
						{"principal" 1280.78M, "start" "2016-07-13"}]
			})
			response (request-application "/debts" json-request)]

			(is (= 200 (:status response)))
			(is (= "application/json; charset=utf-8" (get-in response [:headers "Content-Type"])))
			(is (= expected-response (:body response)))
		)
	)

	(testing "nonexistent endpoint"
		(let [json-request (generate-string {})
			expected-response "Resource not found"
			response (request-application "/nonexistent" json-request)]

			(is (= 404 (:status response)))
			(is (= "text/html; charset=utf-8" (get-in response [:headers "Content-Type"])))
			(is (= expected-response (:body response)))
		)
	)
)


test/webbank/date_converter_test.clj
(ns webbank.date-converter-test
	(:require
		[clojure.test :refer :all]
		[clj-time.core :as clj-time]
		[webbank.date-converter :as date-converter]))


(def br-valentines-string "2015-06-12")
(def br-valentines-date (clj-time/local-date 2015 06 12))

(def easter-string "2017-04-16")
(def easter-date (clj-time/local-date 2017 04 16))

(def christmas-string "2017-12-25")
(def christmas-date (clj-time/local-date 2017 12 25))

(def reveillon-string "2018-01-01")
(def reveillon-date (clj-time/local-date 2018 01 01))

(deftest test-converter
	(testing "from string to date conversion"
		(is (= br-valentines-date (date-converter/from-string br-valentines-string)))
		(is (= easter-date (date-converter/from-string easter-string)))
		(is (= christmas-date (date-converter/from-string christmas-string)))
		(is (= reveillon-date (date-converter/from-string reveillon-string)))
	)

	(testing "from date to string conversion"
		(is (= br-valentines-string (date-converter/to-string br-valentines-date)))
		(is (= easter-string (date-converter/to-string easter-date)))
		(is (= christmas-string (date-converter/to-string christmas-date)))
		(is (= reveillon-string (date-converter/to-string reveillon-date)))
	)
)

src/webbank/api.clj
(ns webbank.api
  	(:require
  		[ring.util.response :as util]
  		[webbank.bank :as bank]
		[webbank.date-converter :as converter]))


(defn ^:private debug-interceptor []
	;(println (str "Bank status:\n" @bank/bank "\n"))
)


(defn ^:private response
	"Converts map responses to standard json."
	[response-map]
	(debug-interceptor)
	(util/response response-map)
)

(defn request-transaction
	"Forwards http transaction request to bank service."
	[request]
	(let [account-number (get request "account")
		description (get request "description")
		amount (.setScale (new BigDecimal (str (get request "amount"))) 2)
		date (converter/from-string (get request "date"))]
	(bank/add-account-transaction account-number description amount date)
	(response {}))
)

(defn request-balance
	"Forwards http balance request to bank service."
	[request]
	(let [account-number (get request "account")]
	(response
		{"amount" (bank/current-balance account-number)}))
)

(defn request-statement
	"Forwards http statement request to bank service."
	[request]
	(let [account-number (get request "account")
		start-date (converter/from-string (get request "start"))
		end-date (converter/from-string (get request "end"))]
	(response
		(bank/account-statement account-number start-date end-date)))
)

(defn request-debts
	"Forwards http debts request to bank service."
	[request]
	(let [account-number (get request "account")]
	(response
		{"debts" (bank/periods-of-debt account-number)}))
)


src/webbank/bank.clj
(ns webbank.bank
	(:require
		[clj-time.core :as clj-time]
		[webbank.date-converter :as converter]))


(def ^:private empty-bank {})
(def bank (atom empty-bank))


(defn reset-bank
	"Returns bank to its original state."
	[]
	(reset! bank empty-bank)
)

(defn ^:private create-account
	"Adds new empty account to bank identified by account-number."
	[account-number]
	(->
		bank
		(swap! #(assoc % account-number (sorted-map)))
		(get account-number)
	)
)

(defn get-account-by
	"Retrieves bank account identified by account-number. In case there is none, creates a new account and returns it."
	[account-number]
	(if-let [bank-account (get @bank account-number)]
		bank-account
		(create-account account-number)
	)
)

(defn add-account-transaction
	"Adds new bank transaction to account identified by account-number."
	[account-number description amount date]
	(when ((complement zero?) amount)
		(let [new_transaction {:description description :amount amount}]
			(as-> (get-account-by account-number) input
				(get input date [])
				(conj input new_transaction)
				(swap! bank assoc-in [account-number date] input)
			)
		))
)

(defn daily-transactions-balance
	"Returns the balance of all account transactions which happened on date."
	[account date]
	(->>
		(get account date [])
		(map #(:amount %))
		(reduce + 0.00M)
	)
)

(defn account-statement
	"Returns a statement with transactions and balances grouped by date for the bank account identified by account-number."
	([account-number]
	(let [account (get-account-by account-number)]
		(loop [[current-date & next-dates] (keys account)
				statement-map account
				balance-accumulator 0.00M]
			(if current-date
				(do (let [new-balance (+ balance-accumulator (daily-transactions-balance account current-date))
						  current-transactions (get account current-date [])]
					(recur next-dates (assoc statement-map current-date [current-transactions new-balance]) new-balance))
				)
				statement-map
			)
		)
	))

	([account-number start-date end-date]
	(->> (account-statement account-number)
		(drop-while #(clj-time/before? (first %) start-date))
		(take-while #((complement clj-time/after?) (first %) end-date))
		(into (sorted-map))
	))
)

(defn ^:private date-from-daily-statement
	"Extracts transaction date from daily-statement."
	[daily-statement]
	(key daily-statement)
)

(defn ^:private balance-from-daily-statement
	"Extracts balance from daily-statement."
	[daily-statement]
	(if daily-statement
		((comp second val) daily-statement)
		0.00M
	)
)

(defn current-balance
	"Gets current balance related to the bank account identified by account-number."
	[account-number]
	(balance-from-daily-statement (last (account-statement account-number)))
)


(defn ^:private next-transaction-dates
	"Maps a transaction date to their very next for each transaction belonging to the account identified by account-number."
	[account-number]
	(let [transaction-dates (keys (get-account-by account-number))]
		(loop [[current successor :as dates] transaction-dates
				result {}]
			(if (nil? current)
				result
				(recur (rest dates) (assoc result current successor))
			)
		))
)

(defn periods-of-debt
	"Returns the periods where balances of the account (identified by account-number) has been negative."
	[account-number]

	(def eve #(clj-time/minus % (clj-time/days 1)))

	(let [full-statement (account-statement account-number)
		next-transactions (next-transaction-dates account-number)
		periods (for [daily-statement full-statement
					:let [balance (balance-from-daily-statement daily-statement)
						date (date-from-daily-statement daily-statement)
						next-date (get next-transactions date)]
					:when (neg? balance)]
					{
						:principal (.abs balance)
						:start (converter/to-string date)
					  	:end (when next-date (converter/to-string (eve next-date)))
					})
		last-debt (last periods)]
		(if (and last-debt (nil? (:end last-debt)))
			(conj (vec (drop-last periods)) (dissoc last-debt :end))
			(vec periods))
	)
)


src/webbank/core.clj
(ns webbank.core
  	(:require
  		[org.httpkit.server :refer [run-server]]
		[compojure.core :refer :all]
		[compojure.route :as route]
  		[ring.middleware.json :as middleware]
  		[webbank.api :as api]))


(defroutes server-routes
	(POST "/transactions" {body :body} (api/request-transaction body))
	(POST "/balances" {body :body} (api/request-balance body))
	(POST "/statements" {body :body} (api/request-statement body))
	(POST "/debts" {body :body} (api/request-debts body))
	(route/not-found "Resource not found")
)

(defn attach-middleware
	"Adds Ring middlewares to parse json request bodies and setup json response headers."
	[app-routes]
  	(-> app-routes
    	(middleware/wrap-json-body)
    	(middleware/wrap-json-response))
)

(def application (attach-middleware server-routes))

(defn -main
	"Starts web server on port 5000."
	[& args]
	(println "Web server running on port 5000.")
	(run-server application {:port 5000})
)

src/webbank/routes.clj
(ns webbank.routes
  	(:require
		[compojure.core :refer :all]
  		[compojure.route :as route]
  		[ring.util.response :as util]
  		[webbank.bank :as bank]))


(defn response
	"Converts all map responses to standard json."
	([] (response {}))

	([response-map]
	(util/response {:status 200 :headers {"Content-Type" "application/json"} :body response-map}))
)


(defn credit-account
	"Forward http credit request to bank service."
	[request]
	(let [account-number (get request "account")
		description (get request "description")
		amount (get request "amount")
		date (get request "date")]
	(bank/add-account-operation account-number description amount date))
	(response)
)

(defn debit-account
	"Forward http debit request to bank service."
	[request]
	(let [account-number (get request "account")
		description (get request "description")
		amount (get request "amount")
		date (get request "date")]
	(bank/add-account-operation account-number  description (- amount) date))
	(response)
)

(defn request-balance
	"Forward http balance request to bank service."
	[request]
	(let [account-number (get request "account")]
	(response
		{:balance (bank/current-balance account-number)}))
)

(defn request-statement
	"Forward http statement request to bank service."
	[request]
	(let [account-number (get request "account")
		start-date (get request "start")
		end-date (get request "end")]
	(response
		(bank/account-statement account-number start-date end-date)))
)

(defn request-debts
	"Forward http debts request to bank service."
	[request]
	(let [account-number (get request "account")]
	(response
		(bank/periods-of-debt account-number)))
)


(defroutes server-routes
	(POST "/credit" {body :body} (credit-account body))
	(POST "/debit" {body :body} (debit-account body))
	(POST "/balance" {body :body} (request-balance body))
	(POST "/statement" {body :body} (request-statement body))
	(POST "/debts" {body :body} (request-debts body))
	(route/not-found (util/response {:status 404, :message "Page not found"}))
)
